cbor/shared/src/main/scala/fs2/data/cbor/low/internal/ItemParser.scala

```diff
@@ -0,0 +1,378 @@
+          parseKeyValues(chunk, idx, rest, count - 1, chunkAcc)
+        }
+      }
+    }
+
+  private def parseKeyValues[F[_]](chunk: Chunk[Byte], idx: Int, rest: Stream[F, Byte], chunkAcc: List[CborItem])(
+      implicit F: RaiseThrowable[F]): Pull[F, CborItem, Context[F]] =
+    requireOneByte(chunk, idx, rest, chunkAcc, peek = true).flatMap { case (chunk, idx, rest, chunkAcc, byte) =>
+      if (byte == 0xff) {
+        Pull.pure((chunk, idx + 1, rest, CborItem.Break :: chunkAcc))
+      } else {
+        parseValue(chunk, idx, rest, chunkAcc).flatMap { case (chunk, idx, rest, chunkAcc) =>
+          parseKeyValues(chunk, idx, rest, chunkAcc)
+        }
+      }
+    }
+
+  private def parseMap[F[_]](chunk: Chunk[Byte],
+                             idx: Int,
+                             rest: Stream[F, Byte],
+                             additional: Byte,
+                             chunkAcc: List[CborItem])(implicit F: RaiseThrowable[F]): Pull[F, CborItem, Context[F]] =
+    if (additional == 31) {
+      parseKeyValues(chunk, idx, rest, CborItem.StartIndefiniteMap :: chunkAcc)
+    } else {
+      parseInteger(chunk, idx, rest, additional, chunkAcc).flatMap { case (chunk, idx, rest, chunkAcc, bytes) =>
+        val size = asUnsignedNumber(bytes)
+        parseKeyValues(chunk, idx, rest, size, CborItem.StartMap(size) :: chunkAcc)
+      }
+    }
+
+  private def parseTaggedValue[F[_]](chunk: Chunk[Byte],
+                                     idx: Int,
+                                     rest: Stream[F, Byte],
+                                     additional: Byte,
+                                     chunkAcc: List[CborItem])(implicit
+      F: RaiseThrowable[F]): Pull[F, CborItem, Context[F]] =
+    parseInteger(chunk, idx, rest, additional, chunkAcc).flatMap { case (chunk, idx, rest, chunkAcc, bytes) =>
+      val tag = asUnsignedNumber(bytes)
+      parseValue(chunk, idx, rest, CborItem.Tag(tag) :: chunkAcc)
+    }
+
+  private def parseSimple[F[_]](chunk: Chunk[Byte],
+                                idx: Int,
+                                rest: Stream[F, Byte],
+                                additional: Byte,
+                                chunkAcc: List[CborItem])(implicit
+      F: RaiseThrowable[F]): Pull[F, CborItem, Context[F]] =
+    (additional: @switch) match {
+      case 20 =>
+        Pull.pure((chunk, idx, rest, CborItem.False :: chunkAcc))
+      case 21 =>
+        Pull.pure((chunk, idx, rest, CborItem.True :: chunkAcc))
+      case 22 =>
+        Pull.pure((chunk, idx, rest, CborItem.Null :: chunkAcc))
+      case 23 =>
+        Pull.pure((chunk, idx, rest, CborItem.Undefined :: chunkAcc))
+      case 24 =>
+        requireOneByte(chunk, idx, rest, chunkAcc).flatMap { case (chunk, idx, rest, chunkAcc, byte) =>
+          if (byte >= 0 && byte < 32)
+            Pull.raiseError(new CborException(s"invalid simple value additional byte $byte"))
```

Comment:
There's only `CborException` right now, right? Would it maybe make sense to introduce two types to differentiate between errors on the low-level parsing vs errors on the high-level parsing? Not that I can come up with a compelling use case right now, but it wouldn't harm and is hard/impossible to change once bincompat guarantees are given.

Reply:
I am generally speaking not a big fan of exception proliferation. I would see a reason to split between parsing and deserialization errors, but I don't want to create too many of them for parsing.

Reply:
I don't mean one for every possible error, but at least introducing a `sealed trait CborException` already with two subtypes. If we then discover valid use cases for more detailed errors, we can add them within bincompat.

Reply:
Meh, why not.